### PR TITLE
Sniper scope

### PR DESCRIPTION
**Pull request by Badulundabad**
- added sniper_scope.xml;
- added SniperScope class;
- added TORCustomBattleAgentStatCalculateModel to change zoom in custom battle;
- overrided GetMaxCamerZoom in TORAgentStatCalculateModel to change zoom in campaign battle;
- adjusted CustomCrosshairMissionBehavior for sniper scope.

### CODE REVIEW

File: CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs [hunk 1]
```diff
@@ -150,5 +141,37 @@ public override void OnPhotoModeDeactivated()
             base.OnPhotoModeDeactivated();
             _weaponCrosshair.OnPhotoModeDeactivated();
         }
+
+        private bool IsPlayerAimingWithSniperRifle()
+        {
+            return Mission.CameraIsFirstPerson &&
+                   Input.IsKeyDown(InputKey.LeftShift) &&
```
Comment (Z3rca) [inline CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs]: note. we really need to try to implement the custom input system. Having hardcoded input checks will be problematic in future

File: CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs [hunk 2]
```diff
@@ -150,5 +163,44 @@ public override void OnPhotoModeDeactivated()
...
+
+            _weaponCrosshair = new Crosshair(Mission, MissionScreen);
+            _weaponCrosshair.InitializeCrosshair();
+            _sniperScope = new SniperScope();
+            _isActive = true;
+
+            if (Agent.Main.IsAbilityUser() && (_abilityComponent = Agent.Main.GetComponent<AbilityComponent>()) != null)
+            {
+                _missionLogic = Mission.Current.GetMissionBehavior<AbilityManagerMissionLogic>();
+                _abilityComponent.CurrentAbilityChanged += (crosshair) =>
```
Comment (hunharibo) [inline CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs]: You should remove the event listener in the Finalize method.

**Review by hunharibo — APPROVED** [review]
One minor comment, nicely cleaned up otherwise :)

File: CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs (hunk 2 above)
Comment (Badulundabad) [reply CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs]: yeah, you're totally right

File: CSharpSourceCode/Battle/Crosshairs/Crosshair.cs [hunk 3]
```diff
@@ -44,132 +43,119 @@ public void FinalizeCrosshair()
...
             bool flag = true;
             bool isTargetInvalid = false;
             for (int i = 0; i < this._targetGadgetOpacities.Length; i++)
             {
                 this._targetGadgetOpacities[i] = 0.0;
             }
-            if (Mission.Current.Mode != MissionMode.Conversation && Mission.Current.Mode != MissionMode.CutScene && Mission.Current.Mode != MissionMode.Deployment && Mission.Current.MainAgent != null && !_missionScreen.IsViewingCharacter() && !this.IsMissionScreenUsingCustomCamera())
+            _crosshairVM.CrosshairType = BannerlordConfig.CrosshairType;
+            Agent mainAgent = Mission.Current.MainAgent;
+            double num = (double)(_missionScreen.CameraViewAngle * 0.017453292f);
```
Comment (Z3rca) [inline CSharpSourceCode/Battle/Crosshairs/Crosshair.cs]: please use a const for variables to get rid of magic numbers

File: CSharpSourceCode/Battle/Crosshairs/Crosshair.cs [hunk 4]
```diff
@@ -44,132 +43,119 @@ public void FinalizeCrosshair()
...
+            double accuracy = 2.0 * Math.Tan((double)(mainAgent.CurrentAimingError + mainAgent.CurrentAimingTurbulance) * (0.5 / Math.Tan(num * 0.5)));
+            _crosshairVM.SetProperties(accuracy, (double)(1f + (_missionScreen.CombatCamera.HorizontalFov - 1.5707964f) / 1.5707964f));
```
Comment (Z3rca) [inline CSharpSourceCode/Battle/Crosshairs/Crosshair.cs]: same here

**Review by Z3rca — COMMENTED** [review]
mostly okay, just get rid of the magic numbers and replace them with  some const.

File: CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs (hunk 1 above)
Comment (Badulundabad) [reply CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs]: this method was removed, because it's not needed

Comment (Badulundabad) [reply CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs]: but i understand your point, moreover we have a lot of methods with using native Input. Anyway i think we will come to implementation our own input someday

File: CSharpSourceCode/Battle/Crosshairs/Crosshair.cs (hunk 3 above)
Comment (Badulundabad) [reply CSharpSourceCode/Battle/Crosshairs/Crosshair.cs]: it's native method, with unknown logic

File: CSharpSourceCode/Battle/Crosshairs/Crosshair.cs (hunk 4 above)
Comment (Badulundabad) [reply CSharpSourceCode/Battle/Crosshairs/Crosshair.cs]: the same)

**Review by Z3rca — APPROVED** [review]

**Pull request merged**